// enka-network-api/dist/models/character/talents/ElementalBurst.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementalBurst = void 0;
const config_file_js_1 = require("config_file.js");
const Element_1 = require("../../Element");
const UniqueSkill_1 = require("./UniqueSkill");
const ExcelTransformer_1 = require("../../../client/ExcelTransformer");
class ElementalBurst extends UniqueSkill_1.UniqueSkill {
    constructor(data, enka) {
        super(data, enka);
        const json = new config_file_js_1.JsonReader(ExcelTransformer_1.excelJsonOptions, this._data);
        this.costElemType = Element_1.Element.getByElementType(json.getAsString("costElemType"), enka);
        this.costElemVal = json.getAsNumber("costElemVal");
        if (this.costElemVal > 0) {
            const cost = this.costElemVal;
            this.requiredCharge = cost;
            this.maxCharge = cost;
        }
        else {
            this.requiredCharge = json.getAsNumber(enka.cachedAssetsManager.getObjectKeysManager().elementalBurstRequiredKey);
            this.maxCharge = json.getAsNumber(enka.cachedAssetsManager.getObjectKeysManager().elementalBurstMaxChargeKey);
        }
    }
    static getById(id, enka) {
        return new ElementalBurst(this._getJsonObjectById(id, enka), enka);
    }
}
exports.ElementalBurst = ElementalBurst;
