// enka-network-api/dist/models/DetailedGenshinUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DetailedGenshinUser = void 0;
const config_file_js_1 = require("config_file.js");
const Character_1 = require("./character/Character");
const GenshinUser_1 = require("./GenshinUser");
class DetailedGenshinUser extends GenshinUser_1.GenshinUser {
    constructor(data, enka) {
        super(data, enka);
        const json = new config_file_js_1.JsonReader(config_file_js_1.defaultJsonOptions, data);
        this.showCharacterDetails = json.has("avatarInfoList");
        this.characters = json.has("avatarInfoList") ? json.get("avatarInfoList").mapArray((_, p) => new Character_1.Character(p.getAsJsonObject(), enka)) : [];
    }
    toGOOD() {
        return {
            format: "GOOD",
            version: 2,
            source: "yuko1101/enka-network-api",
            characters: this.characters.map(c => c.toGOOD()),
            artifacts: this.characters.flatMap(c => c.artifacts).map(a => a.toGOOD()),
            weapons: this.characters.map(c => c.weapon).map(w => w.toGOOD()),
        };
    }
}
exports.DetailedGenshinUser = DetailedGenshinUser;
